sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2665,6 +2666,38 @@ class Analyzer(
       }
     }
   }
+
+  /**
+   * Replace the [[AssignableCast]] expression by [[Cast]], and throw exceptions if the cast is
+   * not assignable.
+   */
+  object ResolveAssignableCast extends Rule[LogicalPlan] {
+    private def fail(from: Expression, to: DataType, walkedTypePath: Seq[String]) = {
+      val fromStr = from match {
+        case l: LambdaVariable => "array element"
+        case e => e.sql
+      }
+      throw new AnalysisException(s"Cannot assign $fromStr from " +
+        s"${from.dataType.catalogString} to ${to.catalogString}.\n" +
+        "The type path of the target object is:\n" + walkedTypePath.mkString("", "\n", "\n") +
+        "You can add an explicit cast to the input data.")
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p if !p.childrenResolved => p
+      case p if p.resolved => p
+
+      case p => p transformExpressions {
+        case u @ AssignableCast(child, _, _) if !child.resolved => u
+
+        case AssignableCast(child, dataType, walkedTypePath)
+          if !Cast.canAssign(child.dataType, dataType) =>
+          fail(child, dataType, walkedTypePath)
+
+        case AssignableCast(child, dataType, _) => Cast(child, dataType.asNullable)
```

Comment:
We don't need rounding/truncating/overflow checks here for some cases, e.g., int->short, double->float?

Reply:
As per the ANSI SQL, rounding/truncating is allowed.
So here we still convert the assignable writes to cast, the result is null if the conversion is out-of-range. Later on, we can add a configuration mode for throwing exceptions on out-of-range conversion.

Reply:
In the current pr, is the result null if out-of-range cases?
For example, in case of `int->short` casts, it seems `Cast` just returns a weired value for a out-of-range value?;
https://github.com/apache/spark/blob/167fa0402dab0800e9762d63d29bd6d9892fa9a8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L479
```
scala> sql("create table t (s short) using parquet")
scala> sql("insert into t values (int(32768))")
// InsertIntoTable Relation[s#12] parquet, false, false
// +- Project [cast(col1#31 as smallint) AS s#32]
//    +- LocalRelation [col1#31]
scala> sql("select * from t").show
+------+
|     s|
+------+
|-32768|
+------+
```

Reply:
@maropu nice catch. It seems that it is an existing issue in Cast. We can fix it first.

Reply:
Thanks!

Reply:
Created https://issues.apache.org/jira/browse/SPARK-28503 for this.

Reply:
Thanks! Just a note; I'm a bit worried about additional overheads to check valid value ranges inside `Cast` since it is already used in many places. As another option, I think we can wrap `Cast` with a `If` expression to check value ranges, e.g., `IF('value range check', CAST, Literal(null))`.